src/main/java/com/limechain/runtime/WasmSectionUtils.java

```diff
@@ -1,124 +1,82 @@
 
-        int searchedKeyIndex = ByteArrayUtils.indexOf(wasmBinary, searchKey);
-        if (searchedKeyIndex < 0) {
-            throw new WasmRuntimeException("Key not found in runtime code.");
-        }
-
-        WasmCustomSection runtimeApis = findRuntimeApisCustomSection(wasmBinary);
-        WasmCustomSection runtimeVersion = findRuntimeVersionCustomSection(wasmBinary);
+        UnknownCustomSection runtimeApis = (UnknownCustomSection) moduleWithCustomSections
+                .customSection(RUNTIME_APIS_SECTION_NAME);
+        UnknownCustomSection runtimeVersion = (UnknownCustomSection) moduleWithCustomSections
+                .customSection(RUNTIME_VERSION_SECTION_NAME);
 
         // If we're missing any of the two custom sections, fallback to `Core_runtime`
         if (Objects.isNull(runtimeApis) || Objects.isNull(runtimeVersion)) {
             return null;
         }
 
         // Read as many api versions as there are (we don't know the length for some reason).
-        ApiVersions apiVersions = ApiVersions.decodeNoLength(runtimeApis.content());
+        ApiVersions apiVersions = ApiVersions.decodeNoLength(runtimeApis.bytes());
 
         // Construct the runtime version partially uninitialized,
         // and then immediately set its apis we've read from the custom section.
-        RuntimeVersion version = new RuntimeVersionReader().read(new ScaleCodecReader(runtimeVersion.content()));
+        RuntimeVersion version = new RuntimeVersionReader().read(new ScaleCodecReader(runtimeVersion.bytes()));
         version.setApis(apiVersions);
 
         return version;
     }
 
     /**
-     * Finds the "runtime_apis" custom section in the given wasm binary
-     */
-    @Nullable
-    public WasmCustomSection findRuntimeApisCustomSection(byte[] wasmBinary) {
-        return findCustomSection(wasmBinary, RUNTIME_APIS_SECTION_NAME);
-    }
-
-    /**
-     * Finds the "runtime_version" custom section in the given wasm binary
+     * Parses the import section of a wasm blob and extracts the "memory" import if present.
+     * @param wasmBinary the wasm blob
+     * @return the parsed {@link org.wasmer.ImportObject.MemoryImport} if present,
+     * a default one with 24 initial pages otherwise
      */
-    @Nullable
-    public WasmCustomSection findRuntimeVersionCustomSection(byte[] wasmBinary) {
-        return WasmSectionUtils.findCustomSection(wasmBinary, RUNTIME_VERSION_SECTION_NAME);
+    public ImportObject.MemoryImport parseMemoryImportFromBinary(byte[] wasmBinary) {
+        ImportSection importSection = toModuleWithSections(wasmBinary, SectionId.IMPORT).importSection();
+        MemoryImport parsedMemoryImport = (MemoryImport) importSection.stream()
+                .filter(i -> i.name().equals("memory"))
+                .findFirst()
+                .orElse(null);
+        return new ImportObject.MemoryImport("env",
+                parsedMemoryImport != null ? parsedMemoryImport.limits().initialPages() : 24,
```

Comment:
A default value of 24 for initial pages seems like an adequate solution considering the current requirements in various runtimes.

Reply:
I'd suggest putting it in a private final field for others that won't know the context of the number

Reply:
Is failing to parse the memory import from the binary an expected scenario or an exceptional case? I feel like, intuitively, we expect the memory import to always be present in the binary (since it throws if there's a mismatch) and failing to retrieve it should be a rather exceptional case, hence bothering with a default value wouldn't be a concern.
Anyway, I might be wrong. Depends on the answer of the first question. 😄

Reply:
Imo it should always be present as per [doc](https://spec.polkadot.network/chap-state#sect-memory-management):

> This memory should be made available to the Polkadot Runtime for import under the symbol name memory.

However my goal was to keep the current implementation to some extent. The previous flow always provided a memory import (regardless of mismatched with runtime wasm or not), which meant that in the case of a missing `memory` import it should startup (and maybe fail if we don't allocate it enough memory for a startup). Imo we should keep this default import, but maybe notify with a log that such an import was not found in the runtime wasm.

Reply:
To my understanding, the cited doc states that we must, in fact, import "a memory", i.e. we must provide an `ImportObject.MemoryImport` when instantiating the module. What I meant was whether we would ever have to "guess a default number for the memory pages" since it seemed to me that the runtime's expectations will always be stated in the wasm blob. And if we can parse that information, then that's our correct and expected way to go.
As for keeping the current implementation - if it doesn't match the use case, I wouldn't strive to keep it. 😄
Anyways, the main question was about intent and how to best express it.

Reply:
It seems that throwing an exception would be a correct way to handle this case. Cases where memory import is missing from the binary should not happen often(never?), thus if it does there would probably be consequences on all other hosts as well. No need to handle it specifically in out imeplementation.